### DISCUSSION TITLE
error: Uncaught (in promise) Error: String required

### DISCUSSION THREAD

**Discussion by gimer** [getwax/bls-wallet#188] [General]
Hi there, 

when I execute the command:

programs/aggregator.ts

I get this error message:

error: Uncaught (in promise) Error: String required
    at y (https://cdn.esm.sh/v74/bls-wallet-clients@0.5.4-f2bad77/deno/bls-wallet-clients.js:2:173085)
    at E1 (https://cdn.esm.sh/v74/bls-wallet-clients@0.5.4-f2bad77/deno/bls-wallet-clients.js:2:172472)
    at $5 (https://cdn.esm.sh/v74/bls-wallet-clients@0.5.4-f2bad77/deno/bls-wallet-clients.js:2:172796)
    at async app (file:///Users/gimer/Cartesi/Aggregator/bls-wallet/aggregator/src/app/app.ts:21:25)
    at async file:///Users/gimer/Cartesi/Aggregator/bls-wallet/aggregator/programs/aggregator.ts:6:1

Any thoughts? Thank you in advance.

**Comment by jacque006** [comment]
Hey @gimer ,

It looks like your network config, which contains contract deployment details, is missing a value. Did you run through all the steps in the [setup](https://github.com/jzaki/bls-wallet#setup) instructions? Specifically:

```sh
yarn hardhat run scripts/deploy_all.ts --network gethDev
```

Which generates that config for local development, or updating ` ./aggregator/.env`:

```
NETWORK_CONFIG_PATH=../contracts/networks/arbitrum-testnet.json
```

For using the Arbitrum Testnet.

**Comment by gimer** [comment]
Dear Jacob, 

thank you for your quick replay, I did go through all the steps in the setup instructions. I think I solved it, I changed the network .json file:

NETWORK_CONFIG_PATH=../contracts/networks/local.json

However, I'm still having problems at the last step. I got this error:

npm notice === Tarball Details === 
npm notice name:          bls-wallet-clients                      
npm notice version:       0.5.4-cca9755                           
npm notice filename:      bls-wallet-clients-0.5.4-cca9755.tgz    
npm notice package size:  118.3 kB                                
npm notice unpacked size: 778.1 kB                                
npm notice shasum:        27d80ab660b8973339c9b863da49b4b5a20925f9
npm notice integrity:     sha512-X+JpFMCgiZj1l[...]zUN+b5jZwCwoQ==
npm notice total files:   228                                     
npm notice 
npm ERR! code ENEEDAUTH
npm ERR! need auth This command requires you to be logged in to https://registry.yarnpkg.com/
npm ERR! need auth You need to authorize this machine using `npm adduser`

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/gimer/.npm/_logs/2022-04-05T00_16_45_612Z-debug-0.log
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.


I followed the steps to add my user, but still not working. Any ideas?

**Comment by voltrevo** [comment]
@gimer Interesting, it's not expected that there would be any npm packages involved that require special permissions.

Could you clarify what you mean by the 'last step'? I would guess that would be the following but it doesn't seem to make sense given the output you posted:

> In a seperate terminal/shell instance
> ```
> cd ./extension
> yarn run dev:chrome # or dev:firefox, dev:opera
> ```

**Comment by voltrevo** [comment]
Actually... is it this step?

> In ./contracts/clients with your changes:
> ```
> yarn publish-experimental
> ```

If so, that does require special permissions. Can you clarify what your overall goal is? That advice is for bls-wallet developers rather than users. If you're working on a fork you might want to consider forking the `bls-wallet-clients` package as well so that you have permissions to publish variations of it.

**Comment by gimer** [comment]
Dear @voltrevo, thank you for your replay. I was just testing this instruction: yarn publish-experimental, but now I understand that it is just optional and that I need special permissions. I'm not working on a fork, my goal at this point is to understand how the BLS signatures and aggregation work. 

I'm currently working as a Blockchain engineer at Cartesi and one of my assignments is to do research on aggregation. I found this project and I think it is very interesting. So far I installed everything and I'm ready to make some tests. Please let me know if you have any advice.

Thank you in advance!

**Reply by jacque006** [reply]
### Running aggregation tests

- You can run tests for the aggregation service with the instructions at https://github.com/jzaki/bls-wallet/blob/main/aggregator/README.md#testing-1 once you have completed the setup instructions
- For the `bls-wallet-clients` package, you can run tests with:
```sh
# From root of project
$ cd ./contracts/clients
$ yarn && yarn build
$ yarn test
```

Both of test suites have BLS signing and signature + transaction aggregation.

### Places  of note in code

#### aggregation service
- https://github.com/jzaki/bls-wallet/blob/main/aggregator/src/app/AggregationStrategy.ts#L122-L205 : Place where aggregation service aggregates transactions

#### client 
- https://github.com/jzaki/bls-wallet/blob/main/contracts/clients/README.md : Examples of how to use `bls-wallet-clients` npm package
- https://github.com/jzaki/bls-wallet/blob/main/contracts/clients/src/signer/sign.ts : transaction (bundle) signing
- https://github.com/jzaki/bls-wallet/blob/main/contracts/clients/src/signer/aggregate.ts : bundle aggregation

#### @thehubbleproject/bls
- https://github.com/thehubbleproject/hubble-bls/blob/master/src/mcl.ts#L177-L184 : Where signatures are aggregated using pairing library
- https://github.com/thehubbleproject/hubble-bls/blob/master/test/signer.test.ts#L27-L47 : Test of the above

Let us know if you have any other questions :)

**Reply by gimer** [reply]
Dear @jacque006, thank you very much! I really appreciate your help. I'll take a look on the information you just sent me. Regards!

**Comment by jacque006** [comment]
Hey @gimer , asked our lead to move this to a discussion. I will post some links to help you out :)

**Reply by gimer** [reply]
Thank you!
